interp/iostream.go

```diff
@@ -0,0 +1,209 @@
+package interp
+
+// I/O streams are interfaces which allow file redirects and command pipelines to be treated
+// equivalently.
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"os/exec"
+	"syscall"
+)
+
+const (
+	notClosedExitCode = -127.0
+)
+
+var (
+	doubleCloseError = errors.New("close: stream already closed")
+)
+
+// firstError returns the first non-nill error or nil if all errors are nil.
+func firstError(errs ...error) error {
+	for _, err := range errs {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Close the cmd and convert the error result into the result returned from goawk builtin functions.
+// A nil error is returned if that error describes a non-zero exit status or an unhandled signal.
+// Any other type of error returns -1.0 and err.
+//
+// The result mimicks gawk for expected child process errors:
+// 1. Returns the exit status of the child process and nil error on normal process exit.
+// 2. Returns 256 + signal on unhandled signal exit.
+// 3. Returns 512 + signal on unhandled signal exit which caused a core dump.
+func waitExitCode(cmd *exec.Cmd) (float64, error) {
+	err := cmd.Wait()
+	if err == nil {
+		return 0.0, nil
+	}
+	ee, ok := err.(*exec.ExitError)
+	if !ok {
+		// Wait() returned an io error.
+		return -1.0, err
+	}
+	status, ok := ee.ProcessState.Sys().(syscall.WaitStatus)
+	if !ok {
+		// Maybe not all platforms support WaitStatus?
+		return -1.0, err
+	}
+	switch {
+	case status.CoreDump():
+		return 512.0 + float64(status.Signal()), nil
+	case status.Signaled():
+		return 256.0 + float64(status.Signal()), nil
+	case status.Exited():
+		return float64(status.ExitStatus()), nil
+	default:
+		return -1.0, err
+	}
+}
+
+type inputStream interface {
+	io.ReadCloser
+	ExitCode() float64
```

Comment:
I think it's a bit weird that these low-level interfaces are dealing in float64s (that's a higher-level AWK value thing). Let's change `iostream.go` to deal in `int` exit codes, and only convert to `float64` where needed in the callers (even if it means a few more type conversions there).

Reply:
Ok sure. But keep in mind the exit codes themselves are also the expected higher-level awk values.